File: Lab1/Lab1/T2.cs
using System.Threading;

namespace Lab1
{
    class T2
    {
        public static void Run(Data data)
        {
            data.print_message(Thread.CurrentThread.Name, "started");

            data.D = data.vector(data.isRandom);                    // 1
            data.MM = data.matrix(data.isRandom);
            //data.print_vector("vector D: ", data.D);
            //data.print_matrix("matrix MM: ", data.MM);

            data.sem2.Release(3);                                   // 2-4
            data.sem1.WaitOne();                                    // 5
            data.sem3.WaitOne();                                    // 6
            data.sem4.WaitOne();                                    // 7

            data.mut1.WaitOne();
            int[] D = data.D;                                       // 8
            int[,] ME = data.ME;                                    // 9
            data.mut1.ReleaseMutex();

            data.operation1(D, ME, data.H, data.H * 2);             // 10
            data.print_message(Thread.CurrentThread.Name, "operation 1 complete");

            data.evh1.Set();                                        // 11
            int ai = data.operation4(data.H, data.H * 2);           // 12
            data.print_message(Thread.CurrentThread.Name, "operation 4 complete");

            lock (data._lock)
            {
                data.operation5(ai);                                // 13
            }
            data.print_message(Thread.CurrentThread.Name, "operation 5 complete");

            data.sem6.Release(3);                                   // 14-16
            data.sem5.WaitOne();                                    // 17
            data.sem7.WaitOne();                                    // 18
            data.sem8.WaitOne();                                    // 19

            data.mut2.WaitOne();
            int a = data.a;                                         // 20
            int x = data.x;                                         // 21
            data.mut2.ReleaseMutex();

            data.operation6(a, x, data.H, data.H * 2);              // 22
            data.print_message(Thread.CurrentThread.Name, "operation 6 complete");
            data.bar1.SignalAndWait();                              // 23

        }
    }
}

File: Lab3/Lab3/src/T2.java
public class T2 extends Thread {
    private final Data data;

    public T2(Data data) {
        this.data = data;
    }

    public void run() {
        data.printMessage("T2", "started");

        data.M1.set_Z(Data.vector());                               // 1

        data.M2.SignalIn();                                         // 2-4
        data.M2.WaitIn();                                           // 5-6

        int d2 = data.M1.get_d();                                   // 7
        int[] Z2 = data.M1.get_Z();                                 // 8

        data.M1.operation1(d2, Z2, data.H, data.H*2);               // 9
        data.printMessage("T2", "operation 1 complete");

        data.M3.SignalSort1();                                      // 10

        int a2 = data.M1.operation4(data.H, data.H*2);              // 11
        data.printMessage("T2", "operation 4 complete");

        data.M1.calc_a(a2);                                         // 12
        data.printMessage("T2", "operation 5 complete");

        data.M2.SignalA();                                          // 13-15
        data.M2.WaitA();                                            // 16-18

        int[] A2 = data.M1.get_A();                                 // 19
        int[][] MX2 = data.M1.get_MX();                             // 20
        a2 = data.M1.get_a();                                       // 21

        data.M1.operation6(A2, MX2, a2, data.H, data.H*2);          // 22
        data.printMessage("T2", "operation 6 complete");

        data.M2.WaitOut();                                          // 23-25

        data.printVector("result vector V:", data.M1.get_V());      // 26
    }
}
